relay-quotas/src/global.rs

```diff
@@ -0,0 +1,256 @@
+use std::cmp::Ordering;
+use std::sync::{Arc, Mutex, OnceLock, RwLock};
+
+use relay_common::time::UnixTimestamp;
+use relay_redis::redis::Script;
+use relay_redis::{PooledClient, RedisError};
+
+use crate::RedisQuota;
+
+fn load_global_lua_script() -> &'static Script {
+    static SCRIPT: OnceLock<Script> = OnceLock::new();
+    SCRIPT.get_or_init(|| Script::new(include_str!("global_quota.lua")))
+}
+
+fn current_slot(window: u64) -> usize {
+    UnixTimestamp::now()
+        .as_secs()
+        .checked_div(window)
+        .unwrap_or_default() as usize
+}
+
+/// Counters used as a cache for global quotas.
+///
+/// When we want to ratelimit across all relay-instances, we need to use redis to synchronize.
+/// Calling Redis every time we want to check if an item should be ratelimited would be very expensive,
+/// which is why we have this cache. It works by 'taking' a certain budget from redis, by pre-incrementing
+/// a global counter. We Put the amount we pre-incremented into this local cache and count down until
+/// we have no more budget, then we ask for more from redis. If we find the global counter is above
+/// the quota limit, we will ratelimit the item.
+#[derive(Default)]
+pub struct GlobalRateLimits {
+    counters: RwLock<hashbrown::HashMap<BudgetKey, Arc<Mutex<SlottedCounter>>>>,
+}
+
+impl GlobalRateLimits {
+    /// Returns `true` if the global quota should be ratelimited.
+    ///
+    /// Certain errors can be resolved by syncing to redis, so in those cases
+    /// we try again to decrement the budget after syncing.
+    pub fn is_rate_limited(
+        &self,
+        client: &mut PooledClient,
+        quota: &RedisQuota,
+        quantity: usize,
+    ) -> Result<bool, RedisError> {
+        let Some(limit) = quota.limit else {
+            return Ok(false);
+        };
+
+        let key = BudgetKeyRef::new(quota);
+
+        let opt_val = self.counters.read().unwrap().get(&key).cloned();
```

Comment:
@jjbayer suggested to use a `Mutex` here since all we do is just a lookup anyways:

```rs
let val = { 
    let counters = self.counters.lock().unwrap_or_else(PoisonError::into_inner);
    Arc::clone(counters.entry_ref(&key).or_default())
};
```

Reply:
even though it's pretty short isn't it still better to not block other threads while we get the counter?